tempesta_fw/pool.c

```diff
@@ -201,6 +211,28 @@ tfw_pool_free(TfwPool *p, void *ptr, size_t n)
 }
 EXPORT_SYMBOL(tfw_pool_free);
 
+/**
+ * Delete all chunks between the last (i.e. current in use) and the first one
+ * (which is the holder of @TfwPool itself). This is a garbage collection
+ * procedure, which is applicable only for cases when pool is used for one
+ * dynamically resizable (via @__tfw_pool_realloc()) instance.
+ */
+void
+tfw_pool_clean(TfwPool *p)
+{
+	TfwPoolChunk *c, *next;
+
+	if (!p)
+		return;
+
+	for (c = p->curr->next; c; c = next) {
+		if (!(next = c->next))
+			break;
+		tfw_pool_free_pages(TFW_POOL_CHUNK_BASE(c), c->order);
```

Comment:
>May cause double-free later at tfw_pool_destroy()
Yeah - missed that. Thanks!
Corrected.

Reply:
Function is used for decoder dynamic table - in `tfw_hpack_add_index()`. Unlike the request/response pools, which live until the message is alive, this pool is intended for decoder table allocations, thus it is tied to the connection life cycle, which is significantly longer than the life period of particular message. In case of increasing decoder table it is worth to free up accumulated unused memory.

Reply:
The idea is clear.
But still that breaks implicit conventions about what pool allocator functions do. The pool allocator was the thing that gives memory chunks that are valid till the allocator instance is valid. So if I saw in the code that memory was allocated and pointer was stored somewhere, I could reason that this memory will be always valid. With the introduction of this clean-up routine it's no longer true. Now even in a code that never will free part of its pools, no one can be sure.

Reply:
>But still that breaks implicit conventions about what pool allocator functions do

I'm not sure that we have such implicit convention in all cases of `TfwPool` usage. E.g. if we use only `__tfw_pool_realloc()` for single potentially large resizable instance - we leave only garbage behind. Specially for this case `tfw_pool_clean()` is created and I left the comment for this procedure with description of the case it is applicable for.

Reply:
I think I understand your point. Introduction of `tfw_pool_clean()` makes it easier to reuse the existing code. It also makes it easier to use pools for the dynamic tables implementation.

But from my perspective that will make code harder to reason about. Previously, it was easy: you create pool, allocate memory, destroy pool. Similar to garbage collection, there were not so much worries about memory life time. Of course, it was possible to destroy a pool, but keep using memory from it, so there were something to keep in mind. However it wasn't possible to free part of pool memory in a sane way, and it made it easier. Now with `tfw_pool_clean()` there is a way. And that change is retroactive, since all code using TfwPool is affected.